### ISSUE TITLE
Not possible to reliably detect a single key press in Lua

### ISSUE THREAD

**Issue by Wuzzy2** [luanti-org/luanti#13029]
## Problem

(This is partially feature request, partially a design flaw maybe.)

There doesn't seem to be a reliable (!) method to detect a single key press.

When I try to detect when the player has pressed a certain key (e.g. Jump), I usually do it like this:

```
-- Will show a chat message while a player presses Jump
-- (might not detect short key press)
minetest.register_globalstep(function(dtime)
   local players = minetest.get_connected_players()
   for p=1, #players do
      local ctrl = players[p]:get_player_controls()
      if ctrl.jump then
          minetest.chat_send_all(players[p]:get_player_name().." has pressed the Jump key!")
      end
   end
end
```

I used this method in Hades Revisited and Glitch and in both games the implementation is not ideal.

So this way to detect input is limited. It only works when I assume the player holds down the key long enough so that `ctrl.jump` will be `true` during a globalstep. But it is possible that if a player pressed the Jump key only shortly, `ctrl.jump` might not be `true` in the globalstep, thus the game misses the input event completely. :-(

I don't know of any other way to get player input right now.

So the current "solution" is to instruct the player to hold down the key longer which isn't great.

## Solution

Add a callback function like `minetest.register_on_player_input(callback, detected_inputs)` with:

* `callback`: A function with parameters `(player)`, called when `player` pressed something
    * `player`: Player object of the player who pressed something
* `detected_inputs`: A table of inputs (same format as `get_player_controls`) where every key with a `true` value will be the keys to look out for. Example: `{jump=true, aux1=true}` will trigger the callback when Jump or Aux1 are pressed and ignores all other keys.

This registers a callback function. The `callback` will be called when a player pressed any of the `detected_keys`.
The callback only fires when the key state goes from "not pressed" to "pressed".
This means it does NOT get called repeatedly while the pressed key is still held down in order to avoid callback spam. When the key was released and is then pressed again, the callback WILL get called again.

### Minimal example

```
-- Will show a chat message whenever a players presses Jump
minetest.register_on_player_input(function(player)
    minetest.chat_send_all(player:get_player_name().." has pressed the Jump key!")
end, {jump=true})
```

**Comment by sfan5** [comment]
The callback you are proposing is close to impossible to make work reliably because it would run on the server, which has a varying amount of latency when talking to the client.
-> (SS)CSM

**Comment by Wuzzy2** [comment]
To be clear what I want:

1. Player presses key
2. Server does something (that only the server can do, like setting nodes)

The important part of my request is that the key press is being detected at all. Latency is allowed.

SSCSMs probably won't help you here because I want the *server* to do something. So there is no way around the latency. And I know the server already is aware of the player input state (otherwise, get_player_controls would not work).

How well aware of the player input state is the server already? Because I wonder if the server actually knows the input state more precisely than mods can detect via `on_globalstep`.

My suggested callback might still be wrong tho, but I don't see how SSCSMs would help either. :-(

**Comment by NathanSalapat** [comment]
I also would like this feature. I'd like to add crouching and crawling to a game, but currently there is no way to sanely let a player hit a key without cycling through multiple poses if they keep the key pressed a fraction of a second too long.

**Comment by sfan5** [comment]
> SSCSMs probably won't help you here because I want the _server_ to do something.

It would, in a way: with CSM you can detect any combination of keys you want without being subject to latency and then send the server a sign over a mod channel that it should do something.
However: yes, in your case CSM is not immediately useful. (cont 1)

> How well aware of the player input state is the server already? 

When the client clicks on a node, object or on air the server gets the input state more than once each step. Otherwise, only once each step which means your Lua approach is close to pretty good as far as current possibilities go.

　

> I'd like to add crouching and crawling to a game, [...]

(cont 1) This is more what I thought your usecase would be and to this I can only say that you do *not* want to do anything that has to do with input and the player model on the server, even if technically possible. The latency will ruin the experience.

**Comment by Panquesito7** [comment]
This would be awesome. Will this be worked on or is it not supported by developers?

**Comment by Zughy** [comment]
Duplicate of #8058
